aes/src/x86/vaes256/aes128.rs

```diff
@@ -0,0 +1,533 @@
+        "vaesenc ymm4 , ymm4 , ymm0",
+        "vaesenc ymm5 , ymm5 , ymm0",
+        "vaesenc ymm6 , ymm6 , ymm0",
+        "vaesenc ymm7 , ymm7 , ymm0",
+        "vaesenc ymm8 , ymm8 , ymm0",
+        "vaesenc ymm9 , ymm9 , ymm0",
+        "vaesenc ymm10, ymm10, ymm0",
+        "vaesenc ymm11, ymm11, ymm0",
+        "vaesenc ymm12, ymm12, ymm0",
+        "vaesenc ymm13, ymm13, ymm0",
+        "vaesenc ymm14, ymm14, ymm0",
+        "vaesenc ymm15, ymm15, ymm0",
+        // aes-128 round 9 encrypt
+        "vmovdqu ymm0 , [{simd_256_keys} +  9 * 32]",
+        "vaesenc ymm1 , ymm1 , ymm0",
+        "vaesenc ymm2 , ymm2 , ymm0",
+        "vaesenc ymm3 , ymm3 , ymm0",
+        "vaesenc ymm4 , ymm4 , ymm0",
+        "vaesenc ymm5 , ymm5 , ymm0",
+        "vaesenc ymm6 , ymm6 , ymm0",
+        "vaesenc ymm7 , ymm7 , ymm0",
+        "vaesenc ymm8 , ymm8 , ymm0",
+        "vaesenc ymm9 , ymm9 , ymm0",
+        "vaesenc ymm10, ymm10, ymm0",
+        "vaesenc ymm11, ymm11, ymm0",
+        "vaesenc ymm12, ymm12, ymm0",
+        "vaesenc ymm13, ymm13, ymm0",
+        "vaesenc ymm14, ymm14, ymm0",
+        "vaesenc ymm15, ymm15, ymm0",
+        // aes-128 round 10 encrypt
+        "vmovdqu ymm0, [{simd_256_keys} + 10 * 32]",
+        "vaesenclast ymm1 , ymm1 , ymm0",
+        "vaesenclast ymm2 , ymm2 , ymm0",
+        "vaesenclast ymm3 , ymm3 , ymm0",
+        "vaesenclast ymm4 , ymm4 , ymm0",
+        "vaesenclast ymm5 , ymm5 , ymm0",
+        "vaesenclast ymm6 , ymm6 , ymm0",
+        "vaesenclast ymm7 , ymm7 , ymm0",
+        "vaesenclast ymm8 , ymm8 , ymm0",
+        "vaesenclast ymm9 , ymm9 , ymm0",
+        "vaesenclast ymm10, ymm10, ymm0",
+        "vaesenclast ymm11, ymm11, ymm0",
+        "vaesenclast ymm12, ymm12, ymm0",
+        "vaesenclast ymm13, ymm13, ymm0",
+        "vaesenclast ymm14, ymm14, ymm0",
+        "vaesenclast ymm15, ymm15, ymm0",
+        // save cipher-data
+        "vmovdqu [{optr} +  0 * 32], ymm1 ",
+        "vmovdqu [{optr} +  1 * 32], ymm2 ",
+        "vmovdqu [{optr} +  2 * 32], ymm3 ",
+        "vmovdqu [{optr} +  3 * 32], ymm4 ",
+        "vmovdqu [{optr} +  4 * 32], ymm5 ",
+        "vmovdqu [{optr} +  5 * 32], ymm6 ",
+        "vmovdqu [{optr} +  6 * 32], ymm7 ",
+        "vmovdqu [{optr} +  7 * 32], ymm8 ",
+        "vmovdqu [{optr} +  8 * 32], ymm9 ",
+        "vmovdqu [{optr} +  9 * 32], ymm10",
+        "vmovdqu [{optr} + 10 * 32], ymm11",
+        "vmovdqu [{optr} + 11 * 32], ymm12",
+        "vmovdqu [{optr} + 12 * 32], ymm13",
+        "vmovdqu [{optr} + 13 * 32], ymm14",
+        "vmovdqu [{optr} + 14 * 32], ymm15",
+
+        simd_256_keys = in(reg) simd_256_keys.as_ptr(),
+        iptr = in(reg) iptr,
+        optr = in(reg) optr,
+
+        clobber_abi("C"),
+        options(nostack, preserves_flags),
+    };
+
+    let iptr = iptr.cast::<Block>();
+    let optr = optr.cast::<Block>();
+
+    let blocks = InOut::from_raw(iptr.add(30), optr.add(30));
+    crate::x86::ni::aes128::encrypt1(keys, blocks);
```

Comment:
This is not great, I would prefer if we just used 30 blocks.

Reply:
The problem with switching from `Block32` to `Block30` is that it results in more complexity elsewhere and would negatively impact the AVX512 case.

For instance, this function uses `Block64` (up from `Block32` due to a previous request IIRC).

```rust
#[inline]
fn proc_par_blocks(&mut self, blocks: InOut<'_, '_, Block64>) {
    unsafe {
        #[cfg(target_arch = "x86_64")]
        if !cfg!(disable_avx512) && self.avx512f_token.get() && self.vaes_token.get() {
            let simd_512_keys = self.simd_512_keys.get_or_insert_with(|| {
                self::vaes512::$module::parallelize_keys(self.keys)
            });
            self::vaes512::$module::encrypt64(simd_512_keys, blocks);
            return;
        }
        #[cfg(target_arch = "x86_64")]
        if !cfg!(disable_avx256) && self.avx_token.get() && self.vaes_token.get() {
            let simd_256_keys = self.simd_256_keys.get_or_insert_with(|| {
                self::vaes256::$module::parallelize_keys(self.keys)
            });
            let (iptr, optr) = blocks.into_raw();
            let iptr = iptr.cast::<Block32>();
            let optr = optr.cast::<Block32>();
            for i in 0..2 {
                let blocks = InOut::from_raw(iptr.add(i), optr.add(i));
                self::vaes256::$module::encrypt32(self.keys, simd_256_keys, blocks);
            }
            return;
        }
        let (iptr, optr) = blocks.into_raw();
        let iptr = iptr.cast::<Block8>();
        let optr = optr.cast::<Block8>();
        for i in 0..8 {
            let blocks = InOut::from_raw(iptr.add(i), optr.add(i));
            self::ni::$module::encrypt8(self.keys, blocks);
        }
    }
}
```

Changing it to `Block30` is problematic since we are relying on runtime feature detection (thus we can't just have a separate function for the `Block30` case). We would either need to process fewer blocks for the VAES 512 case (which seems probably worse), or do as I have it now.

Note that in the part you pointed out, although it looks non-optimal, it doesn't seem to negatively impact performance according to the benchmarks.

Reply:
Relatedly, this might be another point for considering changing the signature for `proc_par_blocks`, which would also benefit VLA style implementations:

> What might be an appropriate API then would be to add a method to one of the traits which would query the implementation for a "block parallelism hint", which would indicate how many blocks would be optimal to pass to the codec. This hint method would in turn execute some of the RVV instructions to calculate what is the (current) maximum available length for processing.

From the RISC-V PR [here](https://github.com/RustCrypto/block-ciphers/pull/399#issue-2092748334).

Reply:
Oh, I wanted to review these parts code later...

This is not a correct use of the backend system. You should define separate backend types for `vaes256` and `vaes512` modules. You should not mix them like this and you certainly should not change `ParBlocksSize` for the XMM AES-NI. Having separate types will allow you to flexibly change `ParBlocksSize`. The only concern is making single block and key expansion functions code from the `ni` module available for `vaes256` and `vaes512`.

Reply:
Re: VLA support. It's a separate discussion, which probably will require a deeper redesign of our trait APIs.